### PR TITLE
server: handle kill signal during write result to connection

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/44009

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by wshwsh12** [comment]
## Manual test
1. Create table t with one int column, and insert 1-8388608. 
2. set @@global.tidb_enable_tmp_storage_on_oom = off; and set @@global.tidb_mem_quota_query = 5 << 30;
3. Use go-mysql-driver to query the sql "select count(1) from t group by a;", fetch 10000 rows and hang forever.
4. Wait 10 minutes and sql doesn't finished.
5. Create another session and try to kill query. kill tidb query xxxxxxxxx.
6. The memory usage is released. After the client is restored, other sql can still be queried.
image

```
tidb> kill tidb query 2631925768;
Query OK, 0 rows affected (0.00 sec)

tidb> show processlist;
+------------+------+-----------------+------+---------+------+---------------------------------+-----------------------------------+
| Id         | User | Host            | db   | Command | Time | State                           | Info                              |
+------------+------+-----------------+------+---------+------+---------------------------------+-----------------------------------+
| 2631925766 | root | 127.0.0.1:35256 | test | Query   |    0 | autocommit                      | show processlist                  |
| 2631925768 | root | 127.0.0.1:37406 | test | Query   |  516 | autocommit; wait query finished | select count(1) from t group by a |
+------------+------+-----------------+------+---------+------+---------------------------------+-----------------------------------+
2 rows in set (0.00 sec)

File: pkg/util/servermemorylimit/servermemorylimit.go [hunk 1]
```diff
@@ -111,10 +111,17 @@ func killSessIfNeeded(s *sessionToBeKilled, bt uint64, sm util.SessionManager) {
 						zap.String("sql text", fmt.Sprintf("%.100v", info.Info)),
 						zap.Int64("sql memory usage", info.MemTracker.BytesConsumed()))
 					s.lastLogTime = time.Now()
+
+					if seconds := time.Since(s.killStartTime) / time.Second; seconds >= 60 {
+						logutil.BgLogger().Warn(fmt.Sprintf("global memory controller failed to kill the top-consumer in %ds, try to close the executors force", seconds))
```
Comment (XuHuaiyu) [inline pkg/util/servermemorylimit/servermemorylimit.go:116]: Add a comment for this

File: pkg/parser/mysql/const.go [hunk 2]
```diff
@@ -59,6 +59,7 @@ const (
 	ServerStatusMetadataChanged    uint16 = 0x0400
 	ServerStatusWasSlow            uint16 = 0x0800
 	ServerPSOutParams              uint16 = 0x1000
+	ServerStatusWaitQueryFinished  uint16 = 0x2000
```
Comment (XuHuaiyu) [inline pkg/parser/mysql/const.go:62]: This needs to be reviewed by the infra team

File: pkg/session/session.go [hunk 3]
```diff
@@ -1474,14 +1474,15 @@ func (s *session) SetProcessInfo(sql string, t time.Time, command byte, maxExecu
 // UpdateProcessInfo updates the session's process info for the running statement.
 func (s *session) UpdateProcessInfo() {
 	pi := s.ShowProcess()
-	if pi == nil || pi.CurTxnStartTS != 0 {
```
[the remark this answers is not in the document]
Comment (wshwsh12) [reply pkg/session/session.go:1477]: Befor this pr, `UpdateProcessInfo` was only used to update `CurTxnStartTS`,so only check whether CurTxnStartTS != 0. This pr needs to update the `State` in `UpdateProcessInfo`, so the update condition needs to be changed.

File: pkg/executor/adapter.go [hunk 4]
```diff
@@ -156,6 +162,14 @@ func (a *recordSet) Next(ctx context.Context, req *chunk.Chunk) (err error) {
 		err = util2.GetRecoverError(r)
 		logutil.Logger(ctx).Error("execute sql panic", zap.String("sql", a.stmt.GetTextToLog(false)), zap.Stack("stack"))
 	}()
+	a.finishLock.Lock()
+	defer a.finishLock.Unlock()
+	if a.stmt != nil {
+		err = a.stmt.Ctx.GetSessionVars().SQLKiller.HandleSignal()
```
Comment (guo-shaoge) [inline pkg/executor/adapter.go:168]:
```suggestion
		if err: = a.stmt.Ctx.GetSessionVars().SQLKiller.HandleSignal()
```

File: pkg/util/sqlkiller/sqlkiller.go [hunk 5]
```diff
@@ -41,11 +41,47 @@ const (
 type SQLKiller struct {
 	Signal killSignal
 	ConnID uint64
+	Finish func()
+	// InWriteResultSet is used to mark whether the query is calling clientConn.writeResultSet().
+	// If the query is in writeResultSet, and Finish() can acquire rs.finishLock, we can think the query is waiting for network IO.
```
Comment (guo-shaoge) [inline pkg/util/sqlkiller/sqlkiller.go:46]:
```suggestion
	// If the query is in writeResultSet, and Finish() can acquire rs.finishLock, we can assume the query is waiting for network IO.
```

**Comment by guo-shaoge** [comment]
Other lgtm

**Review by guo-shaoge — APPROVED** [review]

File: pkg/util/sqlkiller/sqlkiller.go [hunk 6]
```diff
@@ -41,11 +41,47 @@ const (
...
+	InWriteResultSet atomic.Bool
 }
 
 // SendKillSignal sends a kill signal to the query.
 func (killer *SQLKiller) SendKillSignal(reason killSignal) {
-	atomic.CompareAndSwapUint32(&killer.Signal, 0, reason)
+	if atomic.CompareAndSwapUint32(&killer.Signal, 0, reason) {
+		status := atomic.LoadUint32(&killer.Signal)
+		err := killer.getKillError(status)
+		logutil.BgLogger().Warn("kill query started", zap.Uint64("conn", killer.ConnID), zap.String("reason", err.Error()))
```
Comment (XuHuaiyu) [inline pkg/util/sqlkiller/sqlkiller.go:55]: logutil.BgLogger().Warn("kill query initiated", zap.Uint64("connection ID", killer.ConnID), zap.String("reason", err.Error()))

File: pkg/util/sqlkiller/sqlkiller.go [hunk 7]
```diff
@@ -41,11 +41,47 @@ const (
...
+	case QueryMemoryExceeded:
+		return exeerrors.ErrMemoryExceedForQuery.GenWithStackByArgs(killer.ConnID)
+	case ServerMemoryExceeded:
+		return exeerrors.ErrMemoryExceedForInstance.GenWithStackByArgs(killer.ConnID)
+	}
+	return nil
+}
+
+// FinishResultSet is used to finish the result set.
+// If the cancel signal is received and SQL is waiting for network IO, resource released can be performed first.
```
Comment (XuHuaiyu) [inline pkg/util/sqlkiller/sqlkiller.go:80]: // FinishResultSet is used to close the result set.
// If a kill signal is sent but the SQL query is stuck in the network stack while writing packets to the client,
// encountering some bugs that cause it to hang, or failing to detect the kill signal, we can call Finish to release resources used during the SQL execution process.

File: pkg/server/server.go [hunk 8]
```diff
@@ -907,6 +907,12 @@ func (s *Server) Kill(connectionID uint64, query bool, maxExecutionTime bool) {
 		// Mark the client connection status as WaitShutdown, when clientConn.Run detect
 		// this, it will end the dispatch loop and exit.
 		conn.setStatus(connStatusWaitShutdown)
+		if conn.bufReadConn != nil {
+			// When we try `kill connection` and tidb is stuck in the write packet network stack, we can quickly exit the network stack and end the SQL by setting WriteDeadline.
```
Comment (XuHuaiyu) [inline pkg/server/server.go:911]:
```suggestion
// When attempting to 'kill connection' and TiDB is stuck in the network stack while writing packets,
// we can quickly exit the network stack and terminate the SQL execution by setting WriteDeadline.
```

File: pkg/executor/adapter.go [hunk 9]
```diff
@@ -186,16 +199,27 @@ func (a *recordSet) NewChunk(alloc chunk.Allocator) *chunk.Chunk {
...
+			}
+			if err == nil {
+				err = cteErr
+			}
+			a.executor = nil
+			if a.stmt != nil {
+				status := a.stmt.Ctx.GetSessionVars().SQLKiller.GetKillSignal()
+				inWriteResultSet := a.stmt.Ctx.GetSessionVars().SQLKiller.InWriteResultSet.Load()
+				if status > 0 && inWriteResultSet {
+					logutil.BgLogger().Warn("kill query, this SQL may be stuck in the network I/O stack.", zap.Uint64("conn", a.stmt.Ctx.GetSessionVars().ConnectionID))
```
Comment (XuHuaiyu) [inline pkg/executor/adapter.go:218]:
```suggestion
logutil.BgLogger().Warn("kill query, this SQL might be stuck in the network stack while writing packets to the client.", zap.Uint64("connection ID", a.stmt.Ctx.GetSessionVars().ConnectionID))

```

File: pkg/executor/adapter.go [hunk 10]
```diff
@@ -90,17 +90,23 @@ type processinfoSetter interface {
...
+	executor exec.Executor
+	// `Fields` maybe call after `Close`, and executor will clear in `Close` function, so we need to store the schema in recordSet to avoid null pointer exception.
+	schema     *expression.Schema
 	stmt       *ExecStmt
 	lastErrs   []error
 	txnStartTS uint64
 	once       sync.Once
+	// finishLock is a mutex used to synchronize access to the `Next` and `Finish` function of the adapter.
+	// It ensures that only one goroutine can access to the `Next` and `Finish` function at a time, preventing race conditions.
+	// When we terminate the current SQL externally(e.g. kill query), we may use an additional goroutine to call the Finish function.
```
Comment (XuHuaiyu) [inline pkg/executor/adapter.go:103]:
```suggestion
// finishLock is a mutex used to synchronize access to the `Next` and `Finish` functions of the adapter.
// It ensures that only one goroutine can access the `Next` and `Finish` functions at a time, preventing race conditions.
// When we terminate the current SQL externally (e.g., kill query), an additional goroutine would be used to call the `Finish` function.
```

File: pkg/executor/adapter.go (hunk 9 above)
Comment (XuHuaiyu) [inline pkg/executor/adapter.go:218]: will `kill connection` trigger this log?
if so, how do we distinguish `kill query` and `kill connection` from the log?

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**